### PR TITLE
Docs/cookbook upload

### PR DESCRIPTION
**Pull request by maciejkorsan**

### CODE REVIEW

File: docs/cookbook/web/basics/file-upload.md [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+  * Initiated Syncano project
+
+### Problem to solve
+
+You want to upload file to the database.
+
+### Solution
+
+You'll have to create form with upload field and socket to handle it.
+
```
Comment (mkucharz) [inline docs/cookbook/web/basics/file-upload.md]: Too much empty lines.

File: docs/cookbook/web/basics/file-upload.md [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+#### Create socket
+
+At first you need to create a socket. Go to your project's directory with initiated syncano project and type:
```
Comment (mkucharz) [inline docs/cookbook/web/basics/file-upload.md]: s/syncano/Syncano

File: docs/cookbook/web/basics/file-upload.md [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+```sh
+npx s create application
+```
+
+Syncano has `file` type created for storage. You have to create a data class with this type in your `syncano/application/socket.yml` file.
+
+```
+name: application
```
Comment (mkucharz) [inline docs/cookbook/web/basics/file-upload.md]: Let's have here something more like `documents-socket` or something Socket related. This is not really an app.

File: docs/cookbook/web/basics/file-upload.md [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    - name: file
+      type: file
+
+endpoints:
+  upload: 
+    description: add new file
+    parameters: 
+      file: 
+        type: file
+        description: application info
```
Comment (mkucharz) [inline docs/cookbook/web/basics/file-upload.md]: Lack of consequence - Let's have all the descriptions starting with capital letters.

File: docs/cookbook/web/basics/file-upload.md [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+    response:
+      success: 
+        description: Success
+        example:
+          {
+            "msg": "uploaded ok"
```
Comment (mkucharz) [inline docs/cookbook/web/basics/file-upload.md]: Let's change `msg` to message e.g.: `{"message": "Upload successful"}`.

File: docs/cookbook/web/basics/file-upload.md [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+          }
+      fail:
+        exit_code: 400
+        description: Failed
+        parameters:
+          message:
+            description: Error message
```
Comment (mkucharz) [inline docs/cookbook/web/basics/file-upload.md]: No example?

Comment (mkucharz) [reply docs/cookbook/web/basics/file-upload.md]: In the implementation you have "Something went wrong"

File: docs/cookbook/web/basics/file-upload.md [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+      <button class="submit--js">upload</button>
+    </div> 
+  </form>
+
+```
+
+You have to handle form submit with javascript:
+
+```js
+  import Syncano from '@syncano/client'
```
Comment (mkucharz) [inline docs/cookbook/web/basics/file-upload.md]: Is that kind of imports works in `<script>` tags? or it should be webpacked?

File: docs/cookbook/web/basics/file-upload.md [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+  const submit = document.querySelector(".submit--js");
+
+  submit.addEventListener("click", e => {
+    e.preventDefault();
+    const file = document.querySelector("#upload").files[0];
+    let formData = new FormData(); 
+    formData.append("file", file);
+    formData.append("filetype", file.filetype);
+    formData.append("filename", file.name);
+    console.log(formData)
```
Comment (mkucharz) [inline docs/cookbook/web/basics/file-upload.md]: Should we have this print here? I'm more into having comments with actual content of the formdata if we want to some how show it.

File: docs/cookbook/web/basics/file-upload.md [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+    s.post("application/upload", formData).then(res => {
+      console.log(res);
+    });
+  });
```
Comment (mkucharz) [inline docs/cookbook/web/basics/file-upload.md]: `Standard` linter is not accepting `;`, we use it a lot so maybe also here let's clean it up?

File: docs/cookbook/web/basics/file-upload.md [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+      fail:
+        exit_code: 400
+        description: Failed
+        parameters:
+          message:
+            description: Error message
+
+```
+
+Next you need to install `form-data` package using `npm install -D form-data` inside of your socket directory. After that, in your endpoint (`upload.js`) you can to handle the file using it.
```
Comment (adamwardecki) [inline docs/cookbook/web/basics/file-upload.md]: s/can to handle/can handle

File: docs/cookbook/web/basics/file-upload.md [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+  const {data, response} = new Syncano(ctx)
+
+  if (ctx.args.file) {
+    const form = new FormData() 
+    form.append('file', ctx.args.file, {
+      filename: ctx.args.filename,
+      filetype: ctx.args.filetype
+    })
+    
+    data.files.create(form).then( (res) => {
```
Comment (adamwardecki) [inline docs/cookbook/web/basics/file-upload.md]: seems like there also could be a .catch() clause to handle errors on the Syncano side.

File: docs/cookbook/web/basics/file-upload.md [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+```js
+import Syncano from '@syncano/core'
+import FormData from 'form-data'
+
+export default (ctx) => {
+  const {data, response} = new Syncano(ctx)
+
+  if (ctx.args.file) {
+    const form = new FormData() 
+    form.append('file', ctx.args.file, {
```
Comment (adamwardecki) [inline docs/cookbook/web/basics/file-upload.md]:
```
const {file, filename, filetype} = ctx.args
...

  form.append('file', file, { filename, filetype })
```

**Comment by mkucharz** [comment]
@adamwardecki can you hijack this PR?

**Comment by maciejkorsan** [comment]
It would be nice, because I have no time this week to finish this thing ;)

**Comment by adamwardecki** [comment]
Hijack it is then :)

**Pull request merged**